contributors/design-proposals/api-machinery/customresource-conversion-webhook.md

```diff
@@ -125,21 +125,41 @@ type CustomResourceConversionWebhook {
+Hence the defaulting either implicit or explicit has the potential to break backward compatibility as we have two sets of fields representing the same feature. Refer to [this document](http://bit.ly/k8s-crd-per-version-defaulting) for more details.
 
+Hence the solution is the two sets of fields cannot be set at the same time. These validation rules need to be applied:
+
+* Either top level X or per version X can be set not both. This rule applies to individual X’s not the whole set. E.g. top level Schema can be set while per version subresources are set.
+* Per version X cannot be the same. E.g. if all per version schema are the same, the CRD object will be rejected with an error message asking the user to use the top level schema.
+
+The second validation guarantees a clean moving to CRD-v1 (apiextensions.k8s.io/v1). The desired state of CRD-v1 is to only have per-version X and get rid of the top level X. To achieve that, these are conversion rules (and again these rules apply to individual fields not the whole set):
+
+* v1beta1->v1: If top level X is set in v1beta1, then it will be copied to all versions in v1.
+* v1beta1->v1: If per version X are set in v1beta1, then they will be used for per version X in v1.
+* v1->v1beta1: If all per-version X are the same in v1, they will be copied to top level X in v1beta1
+* v1->v1beta1: Otherwise, they will be used as per version X in v1beta1
+
+There is one corner case that if the user of v1beta1 set all per version X the same, then the conversion above is not round trippable. That's the case the above validation protects from.
 
 ### Support Level
 
 The feature will be alpha in the first implementation and will have a feature gate that is defaulted to false. The roll-back story with a feature gate is much more clear. if we have the features as alpha in kubernetes release Y (>X where the feature is missing) and we make it beta in kubernetes release Z, it is not safe to use the feature and downgrade from Y to X but the feature is alpha in Y which is fine. It is safe to downgrade from Z to Y (given that we enable the feature gate in Y) and that is desirable as the feature is beta in Z.
+On downgrading from a Z to Y, stored CRDs can have per version fields set. While the feature gate can be off on Y (alpha cluster), it is dangerous to disable per-version Schema Validation or Status subresources as it makes the status field mutable and validation on CRs will be disabled. Thus the feature gate in Y only protects adding per version fields not the actual behaviour. Thus if the fearure gate is off in Y:
+
+* None of the per-version X cannot be set on create.
+* None of the per-version X cannot be changed on update unless the update command remove them all.
+* Any update that does not change per-version X (if they exists) are allowed.
```

Comment:
My understanding (of https://github.com/kubernetes/kubernetes/pull/67521#issuecomment-417796881; please correct me if we are pursuing a different strategy) is that we allow updates that do change per-version X's if per-version X's exist.

`Any update that does not change per-version X (if they exists)` sounds like "patching top-level X (when per-version X exists)" to me, which would fail the mutual exclusivity validation.

Reply:
I think when the feature gate is disabled, there should be something wrong with the feature and we should not less user guess what they can do to make it work. Any change to the feature may make it worse. But I don't feel strong about it and can go with your way too (except the auto-clear part).

on the auto-clear part, while we are doing that for other feature gates, I object to doing it here. End user of the cluster does not have access to feature gate and if admin disable a feature gate, Operators that installed before that act differently than operator installed after that and both installation goes successfully. This will just create confusion for people.

I agree to continue with auto-clear for now as we do it elsewhere.